server/lib/s3backup/__init__.py

```diff
@@ -0,0 +1,146 @@
+
+    def put_object(self, Bucket, Key, Body, ContentMD5):
+        return self.s3connector.put_object(Bucket=Bucket, Key=Key,
+                Body=Body, ContentMD5=ContentMD5)
+
+
+class MockS3Connector(object):
+    """
+    The mock object is used for unit testing. It provides a "connector" to the backend service that is implemented using the local filesystem, rather than dispatching to the real S3 backend service.
+    """
+    def __init__(self, path, bucket_name):
+        self.path = path
+        self.bucket_name = bucket_name
+
+    def ts(self, *args):
+        return "1900-01-01T00:00:00-UTC"
+
+    def list_buckets(self):
+        return os.listdir(self.path)
+
+    def list_objects(self, Bucket=None):
+        ob_dict = {}
+        result_list = glob.glob(os.path.join(
+            self.path, self.bucket_name, "*.tar.xz"))
+        result_name_list = []
+        for i in result_list:
+            with open(i, 'rb') as f:
+                data = f.read()
+                md5 = hashlib.md5(data).hexdigest()
+            result_name_list.append({'ETag': '"{}"'.format(md5), 'Key': i})
+        ob_dict['Contents'] = result_name_list
+        ob_dict['ResponseMetadata'] = {'HTTPStatusCode': 400}
+        return ob_dict
+
+    def put_object(self, Bucket=None, Key=None, Body=None, ContentMD5=None):
+        md5_hex_value = hashlib.md5(Body).hexdigest()
+        md5_base64_value = (base64.b64encode(
+                        bytes.fromhex(md5_hex_value))).decode()
+        if md5_base64_value == ContentMD5:
+            with open('{}/{}/{}.tar.xz'.format(self.path, self.bucket_name, Key), 'wb') as f:
+                f.write(Body)
+
+
+class S3Config(object):
+    def __init__(self, config):
+        try:
+            debug_unittest = config.get('pbench-server-backup', 'debug_unittest')
+        except Exception as e:
+            debug_unittest = False
+        else:
+            debug_unittest = bool(debug_unittest)
+
+        self.endpoint_url = config.get('pbench-server-backup', 'endpoint_url')
+        self.bucket_name = config.get('pbench-server-backup', 'bucket_name')
+        self.access_key_id = config.get('pbench-server-backup', 'access_key_id')
+        self.secret_access_key = config.get('pbench-server-backup', 'secret_access_key')
+
+        if debug_unittest:
+            self.connector = MockS3Connector(self.endpoint_url, self.bucket_name)
+        else:
+            botoconnector = boto3.client('s3',
+                                   aws_access_key_id='{}'.format(self.access_key_id),
+                                   aws_secret_access_key='{}'.format(self.secret_access_key),
+                                   endpoint_url='{}'.format(self.endpoint_url))
+            self.connector = S3Connector(botoconnector)
+
+    def ts(self, *args):
+        return self.connector.ts(*args)
+
+    def upload_to_bucket(self, tarlist, report, _logger):
+        buck_check = self.connector.list_buckets()
```

Comment:
We are fetching the bucket list since we wanted to make sure beforehand that the bucket should be present for any operation. 

On present code line 121, we are handling an exception where there might be the case that the bucket is present but it is empty. In that case, the `s3_contents` should be an empty list.